### ISSUE TITLE
[Android] when plays mptm tunes (which doesn't supported), and skip to next song, it'll skip some tune after mptm tune.

### ISSUE THREAD

**Issue by knight-ryu12** [libxmp/libxmp#66]
I couldn't provide screenshot. sorry.

**Comment by cmatsuoka** [comment]
Does it happen with any mtpm module?

**Comment by knight-ryu12** [comment]
Yes and No. OpenMPT Exsample song does use outside dir provided in Sample dir.
Saga's Official site released Shooting Star will play fine, But that bug happens.

**Comment by knight-ryu12** [comment]
Refer to all android issue which I made, #68 #63, And this.
I am using Xperia Z2 Tablet, Android 4.4.2.

**Comment by knight-ryu12** [comment]
kinda buffer underrun. (what I hear like)

**Comment by LossyDragon** [comment]
This is still happening with the latest bug fix commits of this post. Saga's Shooting Star is a good example of this issue happening. 

Just playing the song will work fine. But if you press previous to restart the module, click on the module again in the file explorer (while still playing), or go to the previous played module in queue. The song prematurely ends. 

Video: https://lossy.s-ul.eu/mPpQqNUY

I never got around in getting the cli player to compile and test out to see if this is Android specific or not. 

Note: I noticed if you seek around the first couple of times with Shooting Star, the song will freeze. But after that you can seek around just fine.

**Comment by knight-ryu12** [comment]
for I noticed that that MPTM contains some quirks to delay the pattern too much. and seems xmp have issues on that. need more research.

**Comment by LossyDragon** [comment]
I think another issue should be put in for better MPTM support and to address some of the quirks.

As for the skipping. I managed to butcher a windows variant of xmp-cli, I couldn't get the aforementioned song (and a few others) to skip or freeze/stutter as shown in the video. 

Some more investigation is needed to see if its platform related or maybe with the player service getting some call that a new mod suddenly reached end of playback. 🤷

**Comment by LossyDragon** [comment]
So I have basic understanding of what's going on. However; my knowledge of C is not my strong suit. Maybe @AliceLR or @sezero could help me out a little bit since they've been doing an amazing job fixing a whole lot of stuff. 

The issue starts [here](https://github.com/cmatsuoka/xmp-android/blob/8a42a3327d76649d70a7c17c46e926e3275079c3/app/src/main/java/org/helllabs/android/xmp/service/PlayerService.java#L416), and going through all the jni code, it leads to [player.c](https://github.com/libxmp/libxmp/blob/604832e52fce256b14fea1f151a535980c41373c/src/player.c#L1717). 

Skipping through songs, works fine, with some debug logs placed inside `xmp_play_buffer`. I dont see anything out of the ordinary with other modules with play/pause/skip/previous. Though when I press back to play a previous module (while playing the aforementioned song), it suddenly starts returning `-1` at line `1749` at least 22 times based on my placement of debug logs.

 Could it be that the on the android app is polling `xmp_play_buffer` too quickly or is there something else going on?

**Comment by LossyDragon** [comment]
So I rebuilt a new version of libxmp; from master, for the android client and I noticed the freezing/stuttering issue is resolved. It might have been the work being done in #251

**Comment by AliceLR** [comment]
Wonder if that might be due to cleaning up dangling pointers on `xmp_release_module`... all of the other patches for that issue so far have been very specific to their loaders and the one affecting the IT loader was a fairly marginal issue.

**Comment by LossyDragon** [comment]
Maybe! Even when I didn't seed while playing Shooting star, the next song in queue would immediately freeze for a bit. That doesn't seem to be the case anymore. 😄

**Comment by LossyDragon** [comment]
Just an update on this. It seems to happen the most when the playing module is in a pattern delay.

**Comment by AliceLR** [comment]
After digging through the code a bit, I think the bug I found is almost certainly the cause of this issue. My suspicion was right: this is old `f->rowdelay` values leaking through to the next module. This causes `check_end_of_module` to incorrectly detect the end of the module (as `f->end_point` is 1, as the scan `scan_cnt` value for the new module does not expect a pattern delay).

The reason #240 fixed it is because both players always `xmp_set_position` immediately after `xmp_start_player`, which clears `f->rowdelay`. (As of `libxmp_reset_flow` being added in 4.6.2, this is also fixed if `xmp_set_position` is not used.)

Going to go ahead and close this one as fixed.
